JsonSchema/ErrorMessages.cs

```diff
@@ -51,27 +43,63 @@ private static string Get(CultureInfo? culture = null, [CallerMemberName] string
-    public static string ReplaceTokens(this string message, params (string token, object? value)[] parameters)
+	/// <summary>
+	/// Replaces tokens in the form of `[[token]]` with a specified value, serialized as JSON.
+	/// </summary>
+	/// <param name="message">The message template.</param>
+	/// <param name="parameters">
+	/// Tuple of the token name (without brackets) and the value which will replace it.
+	/// </param>
+	/// <returns>The detokenized string.</returns>
+	[RequiresDynamicCode("Calls JsonSerializer.Serialize and might require reflection for token values serialization. Prefer ReplaceTokens that takes TypeInfo parameters for AOT scenarios.")]
+	[RequiresUnreferencedCode("Calls JsonSerializer.Serialize and might require reflection for token values serialization. Prefer ReplaceTokens that takes TypeInfo parameters for AOT scenarios.")]
+	public static string ReplaceTokens(this string message, params (string token, object? value)[] parameters)
     {
         var current = message;
         var values = new object[parameters.Length];
         for (var i = 0; i < parameters.Length; i++)
         {
             var parameter = parameters[i];
-            values[i] = JsonSerializer.Serialize(parameter.value, _serializerOptions);
+            values[i] = JsonSerializer.Serialize(parameter.value);
             current = current.Replace($"[[{parameter.token}]]", $"{{{i}}}");
         }
 
         return string.Format(current, values);
     }
-}
\ No newline at end of file
+
+	/// <summary>
+	/// Replaces tokens in the form of `[[token]]` with a specified value, serialized as JSON.
+	/// </summary>
+	/// <param name="message">The message template.</param>
+	/// <param name="parameters">
+	/// Tuple of the token name (without brackets) and the value and matching JsonTypeInfo which will replace it.
+	/// </param>
+	/// <returns>The detokenized string.</returns>
+	public static string ReplaceTokens(this string message, params (string token, object? value, JsonTypeInfo typeInfo)[] parameters)
+	{
+		var current = message;
+		var values = new object[parameters.Length];
+		for (var i = 0; i < values.Length; i++)
+		{
+			var parameter = parameters[i];
+			values[i] = JsonSerializer.Serialize(parameter.value, parameter.typeInfo);
+			current = current.Replace($"[[{parameter.token}]]", $"{{{i}}}");
+		}
+
+		return string.Format(current, values);
+	}
+
+	/// <summary>
+	/// Helper to make a parameter for ReplaceTokens in a with static type checking.
+	/// </summary>
+	/// <typeparam name="T"></typeparam>
+	/// <param name="token"></param>
+	/// <param name="value"></param>
+	/// <param name="typeInfo"></param>
+	/// <returns></returns>
+	public static (string token, object? value, JsonTypeInfo typeInfo) MakeParam<T>(
```

Comment:
I feel this can still be cleaner.

What if we create `(string token, T value, JsonTypeInfo<T> typeInfo)` as an actual type (e.g. `ErrorMessageToken<T>` with base `ErrorMessageToken`), then we define implicit operators from the tuple to the type?

Then, we update `.ReplaceTokens()` to take `params ErrorMessageToken[]`.  This method should be able to go away.

(I think `ErrorMessageToken` could also be a child class of `ErrorMessages`, so maybe just `Token`?)

Reply:
I went down that path for a bit but with parameter lists it just wasn't matching / hitting the implicit operators I wanted it to. The version here is the best I could come up with. Perhaps you can check out the branch and try playing around with it to see if you can get it to work?

Reply:
I'll have a play with it in a bit.

Reply:
After playing with it, trying to get the implicit cast working (it didn't), I landed on chaining together multiple replacement calls.

```c#
/// <summary>
/// Replaces tokens in the form of `[[token]]` with a specified value, serialized as JSON.
/// </summary>
/// <param name="message">The message template.</param>
/// <param name="name">The token name (without brackets)</param>
/// <param name="value">The value.</param>
/// <param name="typeInfo">A JsonTypeInfo that matches <paramref name="value"/>.</param>
/// <returns>The detokenized string.</returns>
public static string ReplaceToken<T>(this string message, string name, T? value, JsonTypeInfo<T> typeInfo)
{
	var stringValue = JsonSerializer.Serialize(value, typeInfo!);
	return message.Replace($"[[{name}]]", stringValue);
}
```

Usage looks like this:

```c#
evaluation.Results.Fail(Name, "Item at index [[index]] does not have a value at [[pointer]]"
    .ReplaceToken("index", i - 1, ArrayExtSerializerContext.Default.Int32)
    .ReplaceToken("pointer", specifier.By, ArrayExtSerializerContext.Default.JsonPointer)
);
```

Reply:
ReplaceTokens and the params overload of Fail are gone, ReplaceToken is the replacement. Take a look and see what you think.